### PR TITLE
Introduction of break in RAM Operations

### PR DESCRIPTION
**Pull request by b-scholz**
A break-operation permits terminating loops early knowing that the nullary head is already saturated (i.e. set to true). 

This change request pushes optimisations realized in the Synthesiser to RAM so that the interpreter (LVM & RAMI) can profit from this optimization.

### CODE REVIEW

**Comment by mmcgr** [comment]
Our standard test suite does RAMI + LVM + Synthesised, which is taking us close to the 90 minute timeout (a generous timeout before RAMI was added). The jenkins tests will be a little flaky until we adjust the tests or extend the timeout.

**Comment by b-scholz** [comment]
Could you increase the time-out? Xiaowen will need the RAMI for some time (for his experiments).

**Comment by mmcgr** [comment]
I will, it's just a bit tediously manual so it's not an immediate change.

Note that RAMI is not a problem, it's the increased number of tests introduced with it.

**Comment by mmcgr** [comment]
Jenkins, test this please

File: src/RAMI.cpp [hunk 1]
```diff
@@ -516,11 +518,12 @@ void RAMI::evalOp(const RamOperation& op, const InterpreterContext& args) {
             for (auto ip = range.first; ip != range.second; ++ip) {
                 const RamDomain* data = *(ip);
                 ctxt[scan.getTupleId()] = data;
-                visitSearch(scan);
+                if (!visitSearch(scan)) break;
```
Comment (mmcgr) [inline src/RAMI.cpp]: This is one of the safest one line if statements to have, but even so, for consistency wrap it with braces.

File: src/RAMI.cpp [hunk 2]
```diff
@@ -654,13 +659,13 @@ void RAMI::evalOp(const RamOperation& op, const InterpreterContext& args) {
             // run nested part - using base class visitor
             if (aggregate.getFunction() == souffle::MAX || aggregate.getFunction() == souffle::MIN) {
                 if (res == (aggregate.getFunction() == souffle::MAX ? MIN_RAM_DOMAIN : MAX_RAM_DOMAIN)) {
-                    return;
+                    return true;
```
Comment (mmcgr) [inline src/RAMI.cpp]: This nested if seems unnecessarily confusing. While you're modifying it, maybe change it to
```
if (agg.func == max && res == MIN_RAM_DOMAIN) {
  return true;
}
if (agg.func == min ...
```

File: src/RAMI.cpp [hunk 3]
```diff
@@ -751,25 +756,35 @@ void RAMI::evalOp(const RamOperation& op, const InterpreterContext& args) {
...
```
Comment (mmcgr) [inline src/RAMI.cpp]: As above, consider improving readability of the ifs

**Pull request merged**